working/1113 - null-asserting compound assignment/proposal.md

```diff
@@ -0,0 +1,160 @@
+
+<postfixOperator> ::= <incrementOperator>  
+```
+
+to
+
+```latex
+<assignmentOperator> ::= `='
+  \alt `!'? <compoundAssignmentOperator>
+  
+<unaryExpression> ::= …
+  \alt <incrementOperator> <assignableExpression> `!'?
+
+<postfixOperator> ::= `!'? <incrementOperator>  
+```
+
+This allows one optional `!` exactly where it affects a composite operation, and nowhere else.
+
+This grammar is not ambiguous. The `++x!` looks like it could be a problem,
+but because prefix-`++` only applies to an `<assignableExpression>`, a
+ny existing valid expression starting with `++x!` would need to be followed 
+by a selector making it into an assignable expression, like `++x!.foo`,
+and because postfix selectors bind stronger than prefix operators,
+there is no valid parse which accepts `++x!` as an unary expression
+when followed by another selector like`.foo`.
+
+### Semantics
+
+For each LHS, the semantics of an expression of the form 
+``lhs `!' op= expr`` is is specified in almost the same way
+as `lhs op= expr` except that:
+
+* After *reading* `lhs`, the operation throws if the value is `null` in the same way as `lhs!` would.
+* The static analysis uses the **NON_NULL** type of the static type of `lhs` for the operator lookup.
+
+For each LHS, an expression of the form ``lhs `!' incrementOp`` is specified
+in almost the same ways as `lhs incrementOp`, except that:
+
+* After *reading* `lhs`, the operation throws if the value is `null` in the same way as `lhs!`.
+* The static analysis uses the **NON_NULL** type of the static type of `lhs` for the `+` operator lookup.
+
+For each LHS, an expression of the form ``incrementOp lhs `!'`` is specified as behaving the same as `lhs! += 1 `.
+
+### Example
+
+The <code>*e*<sub>1</sub>.*v*! *op*= *e*<sub>2</sub></code> case would be specified as:
+
+```latex
+\LMHash{}%
+\Case{\code{$e_1$.$v$! $op$= $e_2$}}
+Consider a compound assignment $a$ of the form \code{$e_1$.$v$! $op$= $e_2$}.
+Let $x$ be a fresh variable whose static type is the static type of $e_1$.
+Except for errors inside $e_1$ and references to the name $x$,
+exactly the same compile-time errors that would be caused by
+\code{$x$.$v$ = $x$.$v$! $op$ $e_2$}
+are also generated in the case of $a$.
+The static type of $a$ is the static type of \code{$e_1$.$v$! $op$ $e_2$}.
+```
+
+That is, almost exactly the same as for
+<code>*e*<sub>1</sub>.*v* *op*= *e*<sub>2</sub></code>,
+just with a `!`s inserted on all reads of the LHS.
+
+## Possible extensions
+
+### Null checking compound operators
+
+This proposal's `!` modifier handles the case where we *know* that the LHS is non-null.
+We could also introduce a version which checks whether the LHS is `null`,
+and does nothing if it is, similar to the distinction between `?.` and `!.`.
+
+We already have `x ??= y` which does something if `x` is null.
+This one would do something if `x` is non-`null` instead.
+Let's go with one `?` as strawman marker syntax:
+
+```dart
+e.x ?+= 2
```

Comment:
Eh, I think this is a -1 for me. It's starting to get a little too close to operator soup.